File: eulermethods.rb
require 'prime'

class Numeric
  def length
    self.to_s.length
  end  

  def to_a
    self.to_s.split(//).map { |digit| digit.to_i }
  end
  
end

class Fixnum
  def factorial
    factorial = 1
    self.downto(1) {|number| factorial *= number }
    factorial
  end 

  def divisors
    1.upto(Math.sqrt(self)).select {|i| (self % i).zero?}.inject([]) do |f, i| 
      f << i
      f << self / i unless i == self / i
      f
    end.sort
  end

  def sum_divisors
    dvsrs = self.divisors
    dvsrs.delete_at(-1) if dvsrs.count > 1
    dvsrs.sum
  end  

  def permutations
    temp_array = self.to_a.permutation.to_a
    result_array =[]
    temp_array.each do |array|
      result_array << array.join.to_i
    end
    result_array  
  end 

  def abundant?
    self < self.sum_divisors
  end

  def perfect?
    self == self.sum_divisors
  end

  def deficient?
    self > self.sum_divisors
  end 

  def lychrel?
    test_array = []
    temp_value = self
    until test_array.count == 50 || (test_array.count > 0 && test_array[-1].palindrome?)
      temp_value = temp_value + temp_value.reverse
      test_array << temp_value
      break if temp_value.palindrome?
    end
    !test_array[-1].palindrome?
  end  

  def triangular?
    (Math.sqrt((8 * self) + 1)) % 1 == 0
  end    

  def square?
    (Math.sqrt(self)) % 1 == 0
  end  

  def pentagonal?
    ((Math.sqrt((24 * self) + 1) + 1) / 6) % 1 == 0
  end

  def hexagonal?
    ((Math.sqrt((8 * self) + 1) + 1) /  4) % 1 == 0
  end

  def pandigital?
    comparison = (1..self.length).to_a
    test = self.to_a.sort
    test == comparison
  end

  def composite?
    self.divisors.count > 2
  end 

  def rational_square_root?
    Math.sqrt(self) % 1 == 0
  end
  
  def rational_cube_root?
    Math.cbrt(self) % 1 == 0
  end    

  def next_prime
    result_number = self + 1 if self.even?
    result_number = self + 2 if self.odd?
    until result_number.prime?
      result_number += 2
      result_number += 2 if result_number % 5 == 0
    end
    result_number
  end

  def increasing?
    number_string = self.to_s
    increasing_flag = true
    (0..number_string.length-2).each do |index|
      increasing_flag = false if number_string[index].to_i > number_string[index+1].to_i
    end
    increasing_flag
  end
  
  def decreasing?  
    number_string = self.to_s
    decreasing_flag = true
    (0..number_string.length-2).each do |index|
      decreasing_flag = false if number_string[index].to_i < number_string[index+1].to_i
    end
    decreasing_flag
  end  

  def bouncy?
    !self.increasing? && !self.decreasing?
  end

  def reverse
    self.to_s.reverse.to_i
  end

  def palindrome?
    self == self.reverse
  end
  
  def sum_digits
    self.to_a.sum
  end  

end

class Array
  def sum
    self.inject(:+)
  end

end  

class FiboGenerator
  
  # This generator starts with 0 as the zero-indexed term by default.
  # Initialize it with FiboGenerator.new(1) to use 1 as zero-indexed term.

  attr_reader :zero_index

  def initialize(zero_index=0)
    @fibo_numbers = [0,1] if zero_index == 0
    @fibo_numbers = [1,2] if zero_index == 1
    raise "Zero index must be 0 or 1." if zero_index != 0 && zero_index != 1
  end  

  def next
    @fibo_numbers << @fibo_numbers[-1] + @fibo_numbers[-2]
    @fibo_numbers[-1]
  end  

  def generate(n)
    until @fibo_numbers.count == n
      self.next
    end
    @fibo_numbers
  end    

  def show
    @fibo_numbers
  end  

  def count
    @fibo_numbers.count
  end  

end

File: euler_methods.js
#!/usr/bin/env node

module.exports = {

  len: function(number) {
    return (number + "".length);
  },

  toArray: function(number) {
    function splitNumber(n,list) {
      if (list === undefined) {
        list = [];
      }
      if (n < 10) {
        list.unshift(n);
        return list;
      } else {
        list.unshift(Math.floor(n%10));
        return splitNumber(Math.floor(n/10),list);
      }
    }
    return splitNumber(number);
  },

  pandigital: function(number) {
    return (this.toArray(number).sort().join("") === "0123456789");
  },

  factorial: function(number) {
    function factorialIter(product,counter,max) {
      if (counter > max) {
        return product;
      } else {
        return factorialIter((product*counter),(counter+1),max);
      }
    }
    return factorialIter(1,1,number);
  },

  sumDigits: function(number) {
    function sumList(list,index,sum) {
      if (index >= list.length) {
        return sum;
      } else {
        return sumList(list,index+1,sum+list[index]);
      }
    }
    return sumList(this.toArray(number),0,0);
  },

  findDivisors: function(number) {
    var i;
    var j;
    var divisors = [1];
    number % 2 === 0 ? (i = 2, j = 1) : (i = 3, j = 2);
    for (i; i <= Math.floor(number/2); i += j) {
      if (number % i === 0) { divisors.push(i) };
    }
    divisors.push(number);
    return divisors;
  },

  sumDivisors: function(number) {
    function sumList(list,index,sum) {
      if (index >= list.length) {
        return sum;
      } else {
        return sumList(list,index+1,sum+list[index]);
      }
    }
    return sumList(this.findDivisors(number),0,0);
  },

  prime: function(number) {
    return this.findDivisors(number).length === 2 ? true : false;
  },

  composite: function(number) {
    return !this.prime(number)
  },

  abundant: function(number) {
    return number < this.sumDivisors(number);
  },

  perfect: function(number) {
    return number === this.sumDivisors(number);
  },

  deficient: function(number) {
    return self > this.sumDivisors(number);
  },

  triangular: function(number) {
    return (Math.sqrt((8 * number) + 1)) % 1 === 0;
  },

  square: function(number) {
    return (Math.sqrt(number)) % 1 === 0;
  },

  pentagonal: function(number) {
    return ((Math.sqrt((24 * number) + 1) + 1) / 6) % 1 === 0;
  },

  hexagonal: function(number) {
    return ((Math.sqrt((8 * number) + 1) + 1) / 4) % 1 === 0;
  },

  heptagonal: function(number) {
    return (((Math.sqrt((40 * number) + 9)) + 3) / 10) % 1 === 0;
  },

  octagonal: function(number) {
    return (((Math.sqrt((3 * number) + 1)) + 1) / 3) % 1 === 0;
  },

  figurate: function(number) {
    return (this.triangular(number) || this.square(number) ||
      this.pentagonal(number) || this.hexagonal(number) ||
      this.heptagonal(number) || this.octagonal(number));
  },

  increasing: function(number) {
    var digits = this.toArray(number);
    for (var i = 1; i < digits.length; i++) {
      if (digits[i] < digits[i-1]) {
        return false;
      }
    }
    return true;
  },

  decreasing: function(number) {
    var digits = this.toArray(number);
    for (var i = 1; i < digits.length; i++) {
      if (digits[i] > digits[i-1]) {
        return false;
      }
    }
    return true;
  },

  bouncy: function(number) {
    return (!this.increasing(number) && !this.decreasing(number));
  },

  reverse: function(number) {
    return +this.toArray(number).reverse().join("")
  },

  palindrome: function(number) {
    return number === this.reverse(number);
  },

}